src/Symfony/Component/PropertyInfo/Util/PhpDocTypeHelper.php

```diff
@@ -109,19 +110,30 @@ private function createType(DocType $type, bool $nullable, string $docType = nul
         }
 
         if ('[]' === substr($docType, -2)) {
-            if ('mixed[]' === $docType) {
-                $collectionKeyType = null;
+            $collectionKeyType = new Type(Type::BUILTIN_TYPE_INT);
+            $collectionValueType = $this->createType($type, false, substr($docType, 0, -2));
+
+            return new Type(Type::BUILTIN_TYPE_ARRAY, $nullable, null, true, $collectionKeyType, $collectionValueType);
+        }
+
+        if ('array<' === substr($docType, 0, 6) && $type instanceof Array_) {
+            // array<value> is converted to x[] which is handled above
+            // so it's only necessary to handle array<key, value> here
+            $collectionKeyType = $this->getTypes($type->getKeyType())[0];
+
+            // TODO as far as I can see there is no way to define union types yet?
+            $collectionValueTypes = $this->getTypes($type->getValueType());
+            if (\count($collectionValueTypes) > 1) {
                 $collectionValueType = null;
             } else {
-                $collectionKeyType = new Type(Type::BUILTIN_TYPE_INT);
-                $collectionValueType = $this->createType($type, false, substr($docType, 0, -2));
+                $collectionValueType = $collectionValueTypes[0];
```

Comment:
As already stated in the PR comment:

What I noticed while fixing this, is that union types in arrays are not supported in general at the moment, because the Type class supports only one collectionValueType and I do not see how to pass a Union Type there. But I guess that's a different issue and for those types I decided to return null as collection value type for now. (Or better throw on exception?)

Reply:
I will look into this, nullable could work I think. What kind of test do you mean? A failing/incomplete one that needs to be implemented later?